### PR TITLE
Autentification

### PR DESCRIPTION
**Pull request by facucompe**
# Sumary
* Tokens expirate
* Renovation of tokens
* Invalidation of tokens
# Trello Card 
https://trello.com/c/Pcrqqg1x/5-autenticacion
# Screenshot
image
image
image

### CODE REVIEW

**Comment by mishuagopian** [comment]
@facucompe reassign this once you fix my comments in https://github.com/wolox-training/fc-express-js/pull/4. Also, check for conflicts!

File: app/services/sessionManager.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+const jwt = require('jwt-simple'),
+  bcrypt = require('bcrypt'),
+  moment = require('moment'),
+  config = require('../../config/config').config;
+
+const secret = config.common.session.secret;
+
+const maximunDays = 30;
```
Comment (mishuagopian) [inline app/services/sessionManager.js]: typo `n` ---> `maximumDays`

File: app/controllers/users.js [hunk 2]
```diff
@@ -27,17 +34,49 @@ exports.singIn = (req, res, next) => {
...
+exports.renew = (req, res, next) => {
+
+  const accessToken = sessionManager.decode(req.headers.authorization);
+
+  const renewId = req.body ? req.body.renew_id : null;
+
+  if (renewId === accessToken.renewId) {
+    res.status(200);
+    res.send({
+      access_token: sessionManager.createTokenWithRenewId(accessToken.renewId, accessToken.id, accessToken.verificationCode)
```
Comment (mishuagopian) [inline app/controllers/users.js]: `sessionManager.createTokenWithRenewId` only receives 2 params

File: app/controllers/users.js [hunk 3]
```diff
@@ -27,17 +34,49 @@ exports.singIn = (req, res, next) => {
...
+  } else {
+    next(errors.invalidRenewId);
+  }
+};
+
+exports.invalidateAll = (req, res, next) => {
+
+  const accessToken = sessionManager.decode(req.headers.authorization);
+
+  userService.obteinUserWhitID(accessToken.id).then((user) => {
```
Comment (mishuagopian) [inline app/controllers/users.js]: `userService.getById`

File: app/errors.js [hunk 4]
```diff
@@ -16,14 +15,12 @@ exports.invalidUser = {
   message: 'Invalid user or password'
 };
 
-exports.inexistingUser = {
```
Comment (mishuagopian) [inline app/errors.js]: Why have you deleted this?

File: app/routes.js [hunk 5]
```diff
@@ -3,4 +3,6 @@ const users = require('./controllers/users');
 exports.init = (app) => {
   app.post('/users', users.create);
   app.post('/users/sessions', users.singIn);
+  app.post('/users/invalidateAll', users.invalidateAll);
+  app.post('/users/renew', users.renew);
```
Comment (mishuagopian) [inline app/routes.js]: You must also verify, for these two endpoints, few things:
- User is sending a token.
- Token has not expired and is not after maximum useful date.
- Token has the same verification id the user has in the db

File: app/services/users.js [hunk 6]
```diff
@@ -23,17 +23,21 @@ const emailFromWolox = (email) => {
...
-  const serchedEmail = user.email;
-  return orm.models.user.findOne({ where: { email: serchedEmail } }).catch((err) => {
-    throw errors.dbError(err.detail);
+exports.get = (user) => {
+  const email = user.email;
+  return orm.models.user.findOne({ where: { email } }).catch((err) => {
+    return Promise.reject(errors.dbError);
+  });
+};
+exports.getWithID = (id) => {
```
Comment (mishuagopian) [inline app/services/users.js]: getById

File: app/services/users.js [hunk 7]
```diff
@@ -23,17 +23,21 @@ const emailFromWolox = (email) => {
...
 };
 exports.create = (user) => {
-  if (!(validatePassword(user.password) && emailFromWolox(user.email))) {
+  if (!emailFromWolox(user.email)) {
     return Promise.reject(errors.invalidUser);
   }
-  user.password = jwt.encode(user.password, secret);
   return orm.models.user.create(user).catch((err) => {
-    throw errors.dbError(err.errors);
+    return Promise.reject(errors.dbError);
```
Comment (mishuagopian) [inline app/services/users.js]: Why this change? It was ok.

File: test/testUsers.js [hunk 8]
```diff
@@ -55,35 +57,52 @@ describe('/users POST', () => {
 });
 
 describe('/users/sessions POST', () => {
-  it('Should return correct status and user', (done) => {
+  it('Should return correct status, user and a token', (done) => {
```
Comment (mishuagopian) [inline test/testUsers.js]: It's not returning the user..

File: test/testUsers.js [hunk 9]
```diff
@@ -55,35 +57,52 @@ describe('/users POST', () => {
...
 });
+describe('/users/renew POST', () => {
+  it('Should return new token', (done) => {
+    chai.request(server)
+      .post('/users/sessions')
+      .send({ email: 'email@wolox.com.ar', password: 'password123' })
+      .then((res) => {
+        const accessToken = res.body.access_token;
+        const renewId = res.body.renew_id;
+        chai.request(server)
```
Comment (mishuagopian) [inline test/testUsers.js]: this should be returned

File: test/testUsers.js [hunk 10]
```diff
@@ -55,35 +57,52 @@ describe('/users POST', () => {
...
+      .send({ email: 'email@wolox.com.ar', password: 'wrongpassword' })
       .catch((err) => {
         err.should.have.status(400);
         err.response.body.should.have.property('error');
       }).then(() => done());
   });
 });
+describe('/users/renew POST', () => {
+  it('Should return new token', (done) => {
+    chai.request(server)
```
Comment (mishuagopian) [inline test/testUsers.js]: You can create a function that does this successful login for you, so you don't repeat it all over your code

File: app/middlewares/auth.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
+const moment = require('moment'),
+  sessionManager = require('./../services/sessionManager'),
+  userService = require('./../services/users');
+
+exports.secure = (req, res, next) => {
+  const accessToken = sessionManager.decode(req.headers.authorization);
```
Comment (mishuagopian) [inline app/middlewares/auth.js]: We must check if accessToken is not null before any other thing

File: app/middlewares/auth.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+  userService.getByID(accessToken.id).then((user) => {
+    if (user.verificationCode === accessToken.verificationCode
+      && moment().isBefore(accessToken.expirationDate)) {
+      req.user = user;
+      req.accessToken = accessToken;
+      next();
+    } else {
+      res.status(400);
+      res.send({ error:'Invalid token' });
```
Comment (mishuagopian) [inline app/middlewares/auth.js]: I'd define a new error and replace these two lines with:
next(errors.invalidToken);

File: app/middlewares/auth.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+const moment = require('moment'),
+  sessionManager = require('./../services/sessionManager'),
+  userService = require('./../services/users');
+
+exports.secure = (req, res, next) => {
+  const accessToken = sessionManager.decode(req.headers.authorization);
+
+  userService.getByID(accessToken.id).then((user) => {
```
Comment (mishuagopian) [inline app/middlewares/auth.js]: We can check for `moment().isAfter(accessToken.expirationDate)` before accessing the db to avoid it if not necessary

File: app/controllers/users.js [hunk 14]
```diff
@@ -48,3 +50,25 @@ exports.singIn = (req, res, next) => {
...
 };
+
+exports.renew = (req, res, next) => {
+
+  const renewId = req.body ? req.body.renew_id : null;
+
+  if (renewId === req.accessToken.renewId && moment().isBefore(req.accessToken.maximumUsefulDate)) {
+    res.status(200);
+    res.send({
+      access_token: sessionManager.createTokenWithRenewId(req.user, req.accessToken.renewId)
```
Comment (mishuagopian) [inline app/controllers/users.js]: This is not ok as we'll renew both dates doing this. We can expose a function that renews ONLY the expiration date. What do you think about it?

File: app/errors.js [hunk 15]
```diff
@@ -19,3 +19,18 @@ exports.dbError = {
...
   message: 'Data base error'
 };
+
+exports.invalidRenewId = {
+  statusCode: 401,
+  message: 'Invalid renew ID'
+};
+
+exports.inexistingUser = {
+  statusCode: 400,
```
Comment (mishuagopian) [inline app/errors.js]: 404

File: app/errors.js [hunk 16]
```diff
@@ -19,3 +19,18 @@ exports.dbError = {
...
+  message: 'Inexisting user'
```
Comment (mishuagopian) [inline app/errors.js]: not found

File: test/testUsers.js [hunk 17]
```diff
@@ -87,3 +91,20 @@ describe('/users/sessions POST', () => {
       }).then(() => done());
   });
 });
+describe('/users/renew POST', () => {
+  it('Should return new token', (done) => {
+    logIn().then((res) => {
```
Comment (mishuagopian) [inline test/testUsers.js]: You can now reuse this `login` function all over the file

File: test/testUsers.js [hunk 18]
```diff
@@ -87,3 +91,20 @@ describe('/users/sessions POST', () => {
...
+      return chai.request(server)
+        .post('/users/renew')
+        .set('authorization', accessToken)
+        .send({ renew_id: renewId })
+        .then((res2) => {
+          res2.should.have.status(200);
+          res2.body.should.have.property('access_token');
+        });
+    })
+      .then(() => done());
```
Comment (mishuagopian) [inline test/testUsers.js]: Move this to previous line:
```node
}).then(() => done());
```

**Comment by mishuagopian** [comment]
@facucompe few more changes here!

File: app/controllers/users.js [hunk 19]
```diff
@@ -48,3 +50,25 @@ exports.singIn = (req, res, next) => {
     }
   }).catch(next);
 };
+
+exports.renew = (req, res, next) => {
+
+  const renewId = req.body ? req.body.renew_id : null;
+
+  if (renewId === req.accessToken.renewId && moment().isBefore(req.accessToken.maximumUsefulDate)) {
```
Comment (mishuagopian) [inline app/controllers/users.js]: Maybe the `moment().isBefore(req.accessToken.maximumUsefulDate` comparison is no sense as the `expirationDate` will never be bigger than the `maximumUsefulDate` and it will fail when checking for the first one.

Comment (facucompe) [inline app/controllers/users.js]: I'll make this correction in the Game branch
